Added floor implementation in solvers

mijo2 opened the pull request:
#### References to other Issues or PRs
Revives #12296 
Closes #12296 


#### Brief description of what is fixed or changed
--> Added a function that can solve expressions when floor is used in them.

#### Other comments
Example:

```
>>> from sympy import *
>>> from sympy.abc import *
>>> solve(floor(x))
[0, 1]
>>> solve(x*floor(x+1))
[-1, 0]
```

#### Release Notes
<!-- BEGIN RELEASE NOTES -->
* solvers
  * Added a function that can solve expressions when floor is used in them.
<!-- END RELEASE NOTES -->

Commit:
floor() implementation in solve

mijo2 committed:
other: Added PyPI installation procedure

In Installation procedure, details about installing SymPy using
PyPI was added and a minor sentence was fixed.

mijo2 committed:
other: Removed the sudo python install command from README

sudo python setup.py install part was removed from README as it
was bit of an overkill when explaining the basic installation
procedure

Commit:
bug fix

mijo2 committed:
Merge branch 'floor_implementation' of https://github.com/KTaruneshwar/sympy into 12296_floor_implementation

Revived the floor implementation PR

mijo2 committed:
Merge branch 'master' into 12296_floor_implementation

mijo2 commented:
I still have to proper docstrings for ```solve_floor``` function. Will start working on it.

mijo2 commented:
Sorry, accidentally closed the branch.

oscarbenjamin commented on sympy/solvers/solvers.py:
```diff
@@ -1699,7 +1700,7 @@ def _expand(p):
                     soln = list(soln.keys())
 
-                if soln is not None:
+                if soln is not None and not type(gens[0]) == floor:
```
This seems like a fragile method for handling this case. Why in particular should the type not be floor?

mijo2 replied:
I will check and improve this code.

mijo2 replied:
@oscarbenjamin On speculation, we can solve the polynomial when the generator is a ```floor``` but then we would have to check if the ```floor(x)``` is an integer or not. If it is not, then an empty solution will have to be returned. Your thoughts?

ethankward commented on sympy/solvers/tests/test_solvers.py:
```diff
@@ -27,6 +28,12 @@
     return sstr(sympify(e).evalf(n, **options), full_prec=True)
 
+def test_floor():
+    assert solve(floor(x)) == [0, 1]
```
`1` is not a solution to `floor(x) == 0`.

mijo2 replied:
Yes, I will use intervals with right open to give lower and upper limits wherever required.

ethankward replied:
What about when the solution requires multiple intervals? For example, `floor(x^2) = 1`.

mijo2 replied:
For those the current implementation of the function gives incorrect results. I will try to fix them as much as possible.

mijo2 replied:
Since I have revived the PR, I have noticed that certain parts of the code needs to be changed like: ```math.ceil``` has to be removed from the code and the equation that you are suggesting gives correct results. I will make other changes as well and am planning to return a set of Intervals which will mostly be right open or left open wherever required.

mijo2 replied:
@ethankward  In the latest commit, the ```solve_floor``` function solves ```floor(x^2) = 1``` properly and returns answers in terms of Intervals.

mijo2 committed:
solvers: Updated the floor implementation function

Updated the floor implementation function in such a way that now
it returns solutions as a set of Intervals.

mijo2 commented:
@ethankward @oscarbenjamin I have updated the floor implementation function to return answers in terms of Intervals which seems to be appropriate when dealing with ```floor```. I will update this function to later handle nested floor expressions as for now, nested floor expressions is not supported by this current commit update.

ethankward commented:
You need to add a lot more tests for this. As a starting point:

`floor(sin(x))`

`floor(sin(x)) - 1`

`floor(x**3) - 1`

`floor(x**3) - x`

Also wouldn't it make more sense for this to go in `solveset`?

mijo2 commented:
@ethankward ok I will add more test cases to see how well the function works with different expressions along with making it work with nested ```floor``` expressions.

mijo2 commented:
@ethankward since the answers are in sets, then probably it makes sense to add this in solveset but I would see how solveset works to add it there eventually.

oscarbenjamin commented:
I don't think that solve should return intervals.

mijo2 commented:
@oscarbenjamin @ethankward I will move the this functionality to solveset as it seems that it should be there instead in solve

mijo2 commented:
@ethankward I moved the ```floor``` functionality to solveset and it works for algebraic expressions but not with trignometric ones currently. Will update the function according for solving expressions like ```floor(sin(x))```

mijo2 committed:
solvers: Moved the floor function

Moved the floor implementation from solvers to solveset.

mijo2 committed:
solvers: Removed unused imports from solvers and test_solvers

mijo2 commented:
@oscarbenjamin @ethankward I tried solving ```floor(sin(x))``` but it is proving difficult to do so since the answer for ```solveset(sin(x), x)``` is ```Union(ImageSet(Lambda(_n, 2*_n*pi + pi), Integers), ImageSet(Lambda(_n, 2*_n*pi), Integers))``` and answer for ```solveset(sin(x)-1, x)``` is ```ImageSet(Lambda(_n, 2*_n*pi + pi/2), Integers)``` and I don't know how to combine them as of now. Should I add a ```NotImplementedError``` in my code for expressions which involve both ```sin``` and ```floor``` together?

mijo2 commented:
@oscarbenjamin @ethankward When I was going through some limits of this implementation, I found out that a case like: ```floor(x**3) + x - 2``` gives incorrect results. So to counter this case, what we can do is test if ```floor(some function of x)``` and ```x``` appear together. If they do, then we solve the expression by removing the floor and solving, and then intersect the set of solutions with set of integers. How does this solution sound?

ethankward commented:
There are many other examples of periodic functions other than `sin`. Your code should be very specific about which cases it does and doesn't handle. You don't need to solve every single equation, but you should not return incorrect results either.

czgdp1807 commented on sympy/solvers/solvers.py:
```diff
@@ -1539,7 +1539,7 @@ def _solve(f, *symbols, **flags):
 
         # Poly is generally robust enough to convert anything to
-        # a polynomial and tell us the different generators that it
+        # a p   olynomial and tell us the different generators that it
```
Typo?

mijo2 replied:
Yes I plan to fix this with my next update.

mijo2 committed:
solvers: Added recursive check for eliminating expressions not to be evaluated

mijo2 commented:
@ethankward @oscarbenjamin I have added a recursive check for checking for the expressions which shouldn't be evaluated. For now, it doesn't evaluate the Trigonometric and Hyperbolic terms. I will extend this by trying more things out.

mijo2 committed:
solvers: Fixed a typo

ethankward commented:
That approach is not going to work. There are an infinite number of types of equations that you cannot handle other than `TrigonometricFunction` and `HyperbolicFunction`. What about `floor(x) + gamma(x)`? You need to use a positive match rather than a blacklist.

mijo2 commented:
@ethankward so instead of blacklist, I would just check if the expression has only floor type or not. If it does, then the solution will be computed.

mijo2 committed:
solvers: Updated the recursive check in floor expression

Updated the recursive check to not handle expressions when any other
subclass of Function other than floor is in the expression

mijo2 commented:
@ethankward I have updated the implementation, you can take a look. Also, to address ```floor(x) + gamma(x)```, the new update gives the output:
```
NotImplementedError: 
The floor equation cannot be solved with the current implementation.
```

czgdp1807 commented on sympy/solvers/solveset.py:
```diff
@@ -785,6 +787,77 @@ def _solve_radical(f, symbol, solveset_solver):
+    not_compute = False
+    floor_type = False
+
+    if type(f) == floor:
```
```suggestion
    if isinstance(f, floor):
```

czgdp1807 commented:
@mijo2 This has merge conflicts. Please resolve them.

mijo2 commented:
> This has merge conflicts. Please resolve them.

Sure thing. Sorry for the late reply.

mijo2 committed:
Merge branch 'master' into 12296_floor_implementation

mijo2 committed:
Simple code change

Changed an equality expression to ```is_instance``` call.

Co-authored-by: Gagandeep Singh <gdp.1807@gmail.com>
